### DISCUSSION TITLE
AWS X-Ray support?

### DISCUSSION THREAD

**Discussion by aran112000** [brefphp/bref#1053] [Support]
Is there any first party support, or recommended route for using traces in Bref already?

https://github.com/brefphp/bref/blob/573b320d4296c8031c29d60a2788f1f3f56db5c5/src/Context/Context.php#L64

**Comment by mnapoli** [comment]
Nope, unfortunately (as far as I know). Any public content done around X-Ray (sharing experience, documentation, package, …) would be 100% useful because it's missing today.

**Comment by aran112000** [comment]
I've got something working for my initial purpose which I'd be happy to polish up and contribute to introduce first-party Bref support if you're happy with the inclusion? If so, any initial thoughts on how you'd like this as part of Bref? A separate composer package/repo or a PR to be merged into the main project?

**Reply by mnapoli** [reply]
Awesome! It's hard to say for sure without looking at it to be honest: is it a lot of code? Is it low-level (e.g. integrates in the Lambda runtime), or "high-level" (used by userland code directly)? Is it opinionated or close to what AWS provides for other languages?

Having this in a separate package might be the most obvious/natural approach I'd say, but if you have anything online (or some bit of docs or explanation about what it does) that would help.

**Reply by deleugpn** [reply]
I may be wrong but from what I know a separate package is probably what makes most sense. I think there are 3 important factor for X-Ray:

- Trace collection
- Daemon Communication 
- TraceId propagation

For 2), I imagine something like https://github.com/cgauge/php-x-ray-instrumentation/blob/main/src/Daemon.php where we communicate with Lambda X-Ray Native Daemon via UDP.

Item 3) is very tricky as you need to wrap the AWS SDK with adapters.

Lastly, 1) is the most 'spread' part where Symfony would be one way, Laravel another and whatnot.

**Reply by aran112000** [reply]
So a lot of the code we've got so far is pretty application-specific for tracking things like cURL requests, PDO, Redis, Memcached connections/queries... But there is a possible low-level implementation that could exist for others to implement on top of.

For existing packages such as `bref/laravel-bridge` much of this instrumentation could be hooked in to give 'out the box' framework tracing.

With regards to Daemon Communication, what I've got so far is indeed using the Lambda XRay Daemon over UDP to aggregate and push the traces back to AWS. Doing it this way VS the AWS API was substantially faster per request, and means we didn't have to bring through extra AWS lib dependencies.

Thankfully Bref already exposes the UDP address for applications through: `$_SERVER['AWS_XRAY_DAEMON_ADDRESS']`.

As for the TraceIds, these are also exposed by Bref already within the `LAMBDA_INVOCATION_CONTEXT`.

When Tracing is enabled within the serverless.yml, this is when our current logic starts, else it doesn't alter anything:
```yml
provider:
  tracing:
    lambda: true
```

**Reply by deleugpn** [reply]
The item 3) I mentioned is that you have to extract the TraceId from `LAMBDA_INVOCATION_CONTEXT` and then propagate it to SQS, SNS, DynamoDB, Kinesis, S3, etc.

**Reply by aran112000** [reply]
Gotcha. I'm not sure in that example how the trace IDs could be passed along by Bref to those sorts of services as they would normally be invoked within userland code - that would likely need additional time/effort on a project by project basis, no?

**Reply by deleugpn** [reply]
That's what I mean by SDK Wrapper. On NodeJS, AWS provides an X-Ray wrapper themselves so that you can easily swap the SDK with the X-Ray Wrapper for automatic trace propagation. In PHP, nobody is providing the wrapper yet, I think
